ci_tutorial/products_pipeline/full_pipeline.rst

```diff
@@ -0,0 +1,303 @@
+
+.. code-block:: bash
+    :caption: Promoting from products->develop
+
+    # Promotion using Conan download/upload commands 
+    # (slow, can be improved with art:promote custom command)
+    $ conan download --list=uploaded.json -r=products --format=json > promote.json
+    $ conan upload --list=promote.json -r=develop -c
+
+
+And our final ``develop`` repository state will be:
+
+
+.. graphviz::
+    :align: center
+
+    digraph repositories {
+        node [fillcolor="lightskyblue", style=filled, shape=box]
+        rankdir="LR"; 
+        subgraph cluster_0 {
+          label="Packages server";
+          style=filled;
+          color=lightgrey;
+          subgraph cluster_1 {
+            label = "packages\n repository" 
+            shape = "box";
+            style=filled;
+            color=lightblue;
+            "packages" [style=invis];
+            "ai/1.1.0\n (Release)";
+            "ai/1.1.0\n (Debug)";
+          }
+          subgraph cluster_2 {
+            label = "products\n repository" 
+            shape = "box";
+            style=filled;
+            color=lightblue;
+            "products" [style=invis];
+          } 
+          subgraph cluster_3 {
+            rankdir="BT";
+            shape = "box";
+            label = "develop repository";
+            color=lightblue;
+            rankdir="BT";
+    
+            node [fillcolor="lightskyblue", style=filled, shape=box]
+            "game/1.0" -> "engine/1.0" -> "ai/1.0" -> "mathlib/1.0";
+            "engine/1.0" -> "graphics/1.0" -> "mathlib/1.0";
+            "mapviewer/1.0" -> "graphics/1.0";
+            "game/1.0" [fillcolor="lightgreen"];
+            "mapviewer/1.0" [fillcolor="lightgreen"];
+            "ai/promoted" [label="ai/1.1.0\n(new version)"];
+            "engine/promoted" [label="engine/1.0\n(new binary)"];
+            "game/promoted" [label="game/1.0\n(new binary)", fillcolor="lightgreen"];
+            "game/promoted" -> "engine/promoted" -> "ai/promoted" -> "mathlib/1.0";
+             "engine/promoted" -> "graphics/1.0";
+          }
+          {
+            edge[style=invis];
+            "packages" -> "products" -> "game/1.0" ; 
+            rankdir="BT";    
+          }
+        }
+    }
+
+
+This state of the ``develop`` repository will have the following behavior:
+
+- Developers installing ``game/1.0`` or ``engine/1.0`` will by default resolve to latest ``ai/1.1.0`` and use it. They will find pre-compiled binaries for the dependencies too, and they can continue developing using the latest set of dependencies.
+- Developers and CI that were using a lockfile that was locking ``ai/1.0`` version, will still be able to keep working with that dependency without anything breaking, as the new versions and package binaries do not break or invalidate the previous existing binaries.
+
```

Comment:
It'd be nice to have some commentary on what failure modes are likely to arise and how they should be handled.

e.g. what if an intermediate dependency build fails due to one of the changed packages during the product pipeline?

Reply:
@memsharded More specific question/example about rolling out a major version change.

If packages pipeline only use the `develop` repository, and never the `products` repository, how are major versions rolled out?
e.g. if we create a breaking API change in ai/2.0.
Based on my read, this would be the order of events:
- ai/2.0 passes the packages pipeline, using mathlib/1.0 from develop
- ai/2.0 is promoted to the products repository
- products pipeline would run and pass, but not use ai/2.0 b/c engine depends on ai/[>=1.0 <2]
- so ai/2.0 is stuck in products repository until engine updates its version range
- engine updates its version range to ai/[>=2.0 <3], but its package pipeline won't be able to find ai/2.0 b/c it is only looking in develop

It seems like the packages pipeline needs to look at both develop and product repositories so changes like this can work up the chain.

My questions are:
- does this match your experience? should packages pipeline look at both develop and product repositories? (if so, probably good to mention that in the tutorial)
- What order should it look at develop/product repositories? Currently the product pipeline adds product repo first,then develop. This makes sense for the product pipeline that is trying to promote packages from product to develop. However, it may make sense for the packages pipeline to be reversed, first look at the stable develop, second look at the products repo only if a suitable stable package isn't found

Reply:
Very good questions, thanks very much for your feedback @mmmfarrell 

> so ai/2.0 is stuck in products repository until engine updates its version range

Not really!
If you create ``ai/2.0``, and everything is green, the products are ok, etc, because they are not using ``ai/2.0`` the end result is that ``ai/2.0`` gets promoted to the ``develop`` repo. From there it is available for everyone else to keep working downstream to adapt and start using the new ``ai/2.0`` package.
In this case, developers can start working on a ``engine/1.1`` (it is a new version, as it will need source code changes to adapt to the braking ``ai/2.0``), that uses ``ai/2.0``. Only when this new ``engine`` version is green, it will promote to ``develop`` so down to ``game``, the whole app uses the new ``ai/2.0``.

It is true that if a major version happens in a higher package, for example in ``mathlib/2.0``, the CI will need to manage changes from more than 1 PR:
- 1 PR for the ``ai`` package, to use the new ``mathlib/2.0``. It will be blocked in ``products`` pipeline, because the products will break, as ``graphics`` hasn't still upgraded. The packages will be in the ``products`` repos, as they passed correctly the ``packages`` piepline.
- Another PR for the ``graphics`` package, that will also put the binaries in the ``products`` repository.
- Now that both new binaries for both ``ai`` and ``graphics`` have accounted for the changes in ``mathlib/2.0``, now a products pipeline can succeed, and promote both the packages of those PRs that were in the products pipeline.

Basically, devs can work on the different packages PRs, that will accumulate in the ``products`` repository, until they pass all together.

This is exactly the type of discussions and examples that will follow after, the current section is focusing on the most common "happy path" of the CI.

Reply:
I have added some "Final remarks" to the tutorial, to clarify the scope and what will follow in later new sections.